src/pretix/control/forms/event.py

```diff
@@ -502,6 +502,34 @@ def clean(self):
         settings_dict = self.event.settings.freeze()
         settings_dict.update(data)
         validate_event_settings(self.event, data)
+
+        # set all dependants of virtual_keys and
+        # delete all virtual_fields to prevent them from being saved
+        for virtual_key in self.virtual_keys:
+            if virtual_key not in data:
+                continue
+            base_key = virtual_key[:-15]
+            asked_key = base_key + "_asked"
+            required_key = base_key + "_required"
+
+            if data[virtual_key] == 'optional':
+                data[asked_key] = True
+                data[required_key] = False
+            elif data[virtual_key] == 'required':
+                data[asked_key] = True
+                data[required_key] = True
+            # Explicitly check for 'do_not_ask'.
+            # Do not overwrite as default-behaviour when no value for virtual field is transmitted!
+            elif data[virtual_key] == 'do_not_ask':
+                data[asked_key] = False
+                data[required_key] = False
+
+            # hierarkey.forms cannot handle non-existent keys in cleaned_data => do not delete, but set to None
+            data[virtual_key] = None
+
+        # TODO: do we need to re-validate?
+        # if len(self.virtual_keys):
+        #     validate_event_settings(self.event, data)
```

Comment:
Since validate_event_settings can't know anything about "virtual" keys, I think we should **only** validate at this point.

Reply:
virtual_keys is just a list of the keys to access the fields, which are _asked_required and mapped to the _asked and _required. So when the form is submitted, before I loop over all fields listed in virtual_keys, I wanted to have them validated? Probably is not needed as I explicitly check for the allowed values do_not_ask|optional|required, so I will call validate_event_settings after having applied the changes from the virtual_keys-fields.

Reply:
`validate_event_settings` is a pretix-custom function that only exists so we don't need to write the same logic here and in the clean() method of the API. That function only handles edge cases, and will not know how to validate a `_asked_required` key since these only exist in the context of this form, so yeah, validating after sounds better.